Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp

```diff
@@ -375,24 +375,37 @@ void GraphicsContextCG::drawNativeImageInternal(NativeImage& nativeImage, const
     setCGBlendMode(context, options.compositeOperator(), options.blendMode());
 
 #if HAVE(SUPPORT_HDR_DISPLAY)
+    auto oldToneMappingInfo = CGContextGetContentToneMappingInfo(context);
     auto oldHeadroom = CGContextGetEDRTargetHeadroom(context);
 
     auto headroom = options.headroom();
-    auto dynamicRangeLimit = options.dynamicRangeLimit();
-
     if (headroom == Headroom::FromImage)
         headroom = nativeImage.headroom();
 
-    // FIXME: Use CoreGraphics to constrain the brightness of the image more appropriately.
+    CGContextSetEDRTargetHeadroom(context, headroom);
+
     if (headroom > Headroom::None) {
-        static constexpr float maxConstrainedHeadroom = 2;
-        if (dynamicRangeLimit == PlatformDynamicRangeLimit::standard())
-            headroom = Headroom::None;
-        else if (dynamicRangeLimit == PlatformDynamicRangeLimit::constrainedHigh())
-            headroom = std::max<float>(Headroom::None, std::min<float>(headroom * dynamicRangeLimit.value(), maxConstrainedHeadroom));
-    }
+        auto dynamicRangeLimit = options.dynamicRangeLimit();
 
-    CGContextSetEDRTargetHeadroom(context, headroom);
+        float edrStrength = dynamicRangeLimit == PlatformDynamicRangeLimit::noLimit() ? 1 : 0;
+        float cdrStrength = dynamicRangeLimit == PlatformDynamicRangeLimit::constrainedHigh() ? 1 : 0;
+        unsigned averageLightLevel = CGImageGetContentAverageLightLevelNits(subImage.get());
+
+        if (!averageLightLevel)
+            averageLightLevel = CGImageCalculateContentAverageLightLevelNits(subImage.get());
+
+        auto edrStrengthNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberFloatType, &edrStrength));
+        auto cdrStrengthNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberFloatType, &cdrStrength));
+        auto averageLightLevelNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &averageLightLevel));
```

Comment:
Prefer explicit `RetainPtr` over `auto`:
```suggestion
        RetainPtr edrStrengthNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberFloatType, &edrStrength));
        RetainPtr cdrStrengthNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberFloatType, &cdrStrength));
        RetainPtr averageLightLevelNumber = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &averageLightLevel));
```